src/state.cpp

```diff
@@ -153,7 +153,19 @@ State::update(const bytes& leaf_secret)
 MLSPlaintext
 State::remove(LeafIndex removed) const
 {
-  return sign({ Remove{ removed } });
+  uint32_t non_blank_leaves = 0;
+
+  for (uint32_t i = 0; i < _tree.size().val; i++) {
+    const auto& kp = _tree.key_package(LeafIndex{ i });
+    if (kp.has_value()) {
+      if (non_blank_leaves == removed.val) {
+        return sign({ Remove{ LeafIndex{ i } } });
+      }
+      ++non_blank_leaves;
+    }
+  }
+
+  throw InvalidParameterError("Leaf Index mismatch");
```

Comment:
Sorry if I was unclear before.  I think you have the algorithm right now, but I would still prefer things split out.  Something like:

```
struct State {
  struct RosterIndex : public Uint32{};
  // ^^^ suggest putting this here because it's not meaningful without the tree.
  // But you could also make it a stand-alone type, parallel to LeafIndex and NodeIndex.
  // In any case, we should use the type system to avoid confusion here.

  LeafIndex leaf_for_roster_entry(RosterIndex index); // <-- this algorithm
  MLSPlaintext remove(RosterIndex removed) const {
    return remove(leaf_for_roster_entry(removed));
  }
}
```

Reply:
good point .. will do